File: framework/languages/java/e2e/PubSub/Server/Publisher/src/main/java/systems/zlink/e2e/pubsub/publisher/Configuration/PublisherOptions.java
package Configuration;
import systems.zlink.e2e.pubsub.publisher.Configuration;
import systems.zlink.e2e.pubsub.shared.Contracts;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("e2e")
public record PublisherOptions(
    String httpEndpoint,
    String publisherEndpoint,
    Integer publisherPort,
    String routingId,
    String routingIdPrefix,
    String advertiseHost,
    String channelName,
    String redisLocationEndpoint,
    String locationKeyPrefix,
    String logDir) {
    public PublisherOptions {
        required(httpEndpoint, "http-endpoint");
        if ((publisherEndpoint == null || publisherEndpoint.isBlank()) && publisherPort == null) {
            throw new IllegalArgumentException(
                "e2e.publisher-endpoint or e2e.publisher-port is required");
        }
        channelName = channelName == null || channelName.isBlank()
            ? Contracts.EVENT_CHANNEL
            : channelName;
        if (redisLocationEndpoint != null && !redisLocationEndpoint.isBlank()) {
            required(locationKeyPrefix, "location-key-prefix");
        }
        required(logDir, "log-dir");
    }

    private static void required(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("e2e." + name + " is required");
        }
    }
}

File: framework/languages/java/e2e-kotlin/PubSub/Server/Publisher/src/main/kotlin/systems/zlink/e2e/kotlin/pubsub/publisher/Configuration/PublisherOptions.kt
package systems.zlink.e2e.kotlin.pubsub.publisher

data class PublisherOptions(
    val publisherEndpoint: String,
    val httpEndpoint: String,
    val routingId: String?,
    val routingIdPrefix: String?,
    val advertiseHost: String?,
    val channelName: String,
    val listenPort: Int?,
    val redisLocationEndpoint: String?,
    val locationKeyPrefix: String?,
    val logDir: String,
) {
    companion object {
        fun parse(args: Array<String>): PublisherOptions {
            val values = parseArgs(args)
            fun required(key: String): String =
                values[key]?.takeIf { it.isNotBlank() } ?: throw IllegalArgumentException("--$key is required.")
            return PublisherOptions(
                publisherEndpoint = values["publisher-endpoint"]?.takeIf { it.isNotBlank() }.orEmpty(),
                httpEndpoint = required("http-endpoint"),
                routingId = values["rid"]?.takeIf { it.isNotBlank() },
                routingIdPrefix = values["routing-id-prefix"]?.takeIf { it.isNotBlank() },
                advertiseHost = values["advertise-host"]?.takeIf { it.isNotBlank() },
                channelName = values["channel-name"]?.takeIf { it.isNotBlank() }
                    ?: systems.zlink.e2e.kotlin.pubsub.shared.Contracts.EVENT_CHANNEL,
                listenPort = values["publisher-port"]?.takeIf { it.isNotBlank() }?.toInt(),
                redisLocationEndpoint = values["redis-location-endpoint"]?.takeIf { it.isNotBlank() },
                locationKeyPrefix = values["location-key-prefix"]?.takeIf { it.isNotBlank() },
                logDir = values["log-dir"]?.takeIf { it.isNotBlank() } ?: "logs",
            ).also { options ->
                require(options.publisherEndpoint.isNotBlank() || options.listenPort != null) {
                    "--publisher-endpoint or --publisher-port is required."
                }
                require(options.redisLocationEndpoint == null || options.locationKeyPrefix != null) {
                    "--location-key-prefix is required with --redis-location-endpoint."
                }
            }
        }
    }
}

private fun parseArgs(args: Array<String>): Map<String, String> {
    val values = linkedMapOf<String, String>()
    var index = 0
    while (index < args.size) {
        val key = args[index]
        if (!key.startsWith("--")) {
            index++
            continue
        }
        require(index + 1 < args.size) { "Missing value for $key." }
        values[key.removePrefix("--")] = args[index + 1]
        index += 2
    }
    return values
}
